specs/governance/alligator.md

```diff
@@ -0,0 +1,390 @@
+its delegation and checkpoint mappings accordingly. If either address involved in the transfer (`_from_` or `_to`) has not
+been migrated to the `Alligator` contract, the contract copies the address' checkpoint data from the
+`GovernanceToken` to its own state.
+
+## Interface
+
+### Core Functions
+
+#### `subdelegate`
+
+Allows subdelegation of token voting power to another address (delegatee) with a specified subdelegation rule. This function
+is intended to be called by users that require advanced delegation of the `GovernanceToken`.
+
+```solidity
+function subdelegate(address _delegatee, SubdelegationRule _rule) external
+```
+
+This function MUST check if the `msg.sender` or `_delegatee` addresses have been migrated by checking the `migrated` mapping
+from its [storage](#storage). If either address has not been migrated, the `Alligator` MUST copy the delegation
+and checkpoint data from the token contract to its own state. After copying the data, the `Alligator` MUST update
+the `migrated` mapping to reflect that the address has been migrated.
+
+Before updating the subdelegation, the `subdelegate` function MUST check the validity of the subdelegation rule.
+
+When updating the subdelegation, the `subdelegate` function MUST override any previous subdelegation of the `msg.sender`
+to the `_delegatee`. Afterwards, this function MUST emit a `Subdelegation` event with the given function parameters.
+
+#### `subdelegateFromToken`
+
+Allows subdelegation of token voting power from an address (delegator) to another address (delegatee) with a specified
+subdelegation rule. This function is intended to be called by the `GovernanceToken` contract as part
+of its `delegateBySig` function. To ensure backwards compatibility in the `GovernanceToken`, the
+subdelegation rule MUST be 100% delegation, mimicking the behavior of the `GovernanceToken`'s `delegate`
+function.
+
+```solidity
+function subdelegateFromToken(address _delegator, address _delegatee, SubdelegationRule _rule) external
+```
+
+This function MUST check if the `_delegator` or `_delegatee` addresses have been migrated by checking the `migrated` mapping
+from its [storage](#storage). If either address has not been migrated, the `Alligator` MUST copy the delegation
+and checkpoint data from the token contract to its own state. After copying the data, the `Alligator` MUST update
+the `migrated` mapping to reflect that the address has been migrated.
+
+Before updating the subdelegation, the `subdelegateFromToken` function MUST check the validity of the subdelegation rule.
+
+When updating the subdelegation, the `subdelegateFromToken` function MUST override any previous subdelegation of the
+`_delegator` to the `_delegatee`. Afterwards, this function MUST emit a `Subdelegation` event with the given function
+parameters.
+
+#### `subdelegateBatched`
+
+Allows batch subdelegation of token voting power to multiple addresses with specified subdelegation rules. This
```

Comment:
Is the above subdelegation rule 100% delegation to another address? How does this work under this batching function?

Reply:
this function accepts a list of subdelegation rules, which can be of any type. the subdelegation of 100% delegation is only for the `subdelegateFromToken` to mimick the current behavior of delegations in the token.

this batching function will apply each pair of delegatee and subdelegation rule.

Reply:
Got it, so basically you can have any combination of subdelegation you'd like to. In terms of the amount, does it also have to be 100% in the batch, or it can be lower (just for curiosity)